### DISCUSSION TITLE
API v2 (again)

### DISCUSSION THREAD

**Discussion by x1arch** [opensubsonic/open-subsonic-api#48] [General]
I think current API is totally outdated. Now, when I start implement shares and try to make it compatible with current API - it's pain. If this **Open** Subsonic API we can make proposal of new Open Subsonic API version, which cover most of today requests.

I was read discussions:

- https://github.com/opensubsonic/open-subsonic-api/discussions/9
- https://support.symfonium.app/t/subsonic-servers-participation/1233

One of main question how to now which features server support? It is very simple, let's make then server told to us which features it's support. 

I suggest make a new Open Subsonic API, which will implement: 
Method | Req. | Params | Response | Comment
-- | -- | -- | -- | --
POST | /api/auth | { login, password } | { access_token: "access_token", access_token_expires: "date time", renew_token: "renew_token", renew_token_expires: "date time", capabilities: ["list_of_capabilities_here"] } | Auth (access_token will add to head for every request, except public). Capabilities the same as `GET /api/capabilities`.
GET | /api/capabilities | - | { capabilities: ["list_of_capabilities_here"] } | List of community described features (as ex. { capabilities: ["podcast/v1", "share/v1", "feature/v1/subfeature", "proposal/feature", "experimental/feature"]). In documentation we can describe mandatory methods for feature/version and submethods, which show to client more newest methods are supported by server.  Here we can add experimental or proposed capabilities. Of course if client do not support capability, it will ignore it. Capability can be used as URL path to feature.
-- | -- | -- | -- | Server specific methods
GET, POST, PUT, DELETE | /api/srv/... | | | The server specific methods for internal server needs. (as ex. for gonic admin portal or navidrome share methods, etc.)
-- | -- | -- | -- | Public methods (allowed w/o authorization)
GET, POST, PUT | /api/public/... | | | The public methods for shares (maybe server status, etc.).
-- | -- | -- | -- | Playlist example
GET | /api/playlists/v1(?limit=10&skip=200) | | | Return playlist (with pagination, if needed)
GET | /api/playlist/v1/{id} |   |  | Return one playlist by ID
POST | /api/playlist/v1 | { playlist data }  | { playlist response } | Create new Playlist
PUT | /api/playlist/v1/{id} | { changed playlist data } | { playlist response } | Update playlist
DELETE | /api/playlist/v1/{id} | | | Delete playlist 

Of course for answers will use standard HTTP codes:

- 2xx - success
- 4xx - error
- 5xx - for server errors

I suggest make mandatory next features:

- playlists/v1
- albums/v1
- artists/v1

If we wanna add incompatible changes to feature, we will change number of feature version and done. If we wanna make many new changes in feature, we can make it as version update or as subfeature `playlist/v1/awesome_ratings`. When we update feature version, in the documentation we should add `deprecated` notification for old version of feature.

Capabilities is a good way to know what features server are support.

I ready implement this API to gonic / navidrome servers and xonic-ui client very fast.

**Comment by dweymouth** [comment]
We already discussed this at length among server and client developers and the decision (at least for now) was to go with backward compatible API extensions only

**Reply by x1arch** [reply]
I do not tell all should now migrate to API v2, I just wanna start working on that. And devs, which interested in that (as me, @jeffvli, @spl0k, @deluan) can use it for there projects. If you want, you can use old legacy API, but if we want new modern API, why NOT?

PS Other devs will arrive to our party later, if they want.

PPS I do not understand from documentation, how works [extensions](https://opensubsonic.netlify.app/docs/extensions/template/) is it mean extensions died not being born?

**Reply by Tolriq** [reply]
There's a year of discussion and hard work to reach a consensus to have everyone working on the same direction in a proper way.

You can build your own API in your own yard if you want and if you can convince all the servers and clients to migrate to something completely different in a single shot and build an eco system around it then congratz but this is not what this project is about so you'll start another one.

You should take a little time to understand the project and the documentation that explains everything about how OpenSubsonic works.

https://opensubsonic.netlify.app/docs/opensubsonic-changes/

**Reply by epoupon** [reply]
Indeed we decided to go for a backward compatible API. But I do agree it should not prevent us from thinking about a brand new API (edit: but this is another project, not related to this one)

I think the main blocker to move forward for the backward compatible API is to define a reliable, secure authentication scheme

**Reply by Tolriq** [reply]
I know the concept of the extensions is still mysterious for some, but we can build a new API on top of this project this not incompatible, this is just the result of migrating all endpoints to new extensions.

In the end it's the same story, if you only get 1 or 2 server and 1 or 2 client then this reach no purpose and each server can and should do it's own API that better fit it's needs and try to convince clients to adopt it.

Many clients are dying, Navidrome have the large majority of server shares for now, I could have chosen the simple solution to work with @deluan to finish his internal API and have a huge advantage via features others can't provide.
Instead I took tons of time and energy to try to bring this project to life to have a sane ecosystem for servers and clients and most importantly for the users who have choice and choices brings innovations.

Everything he describe in the post above is already covered in the work from for OpenSubsonic. For the auth the issue is not even about having something backward compatible it's about choosing the solution we want. Whatever pre auth system we choose we can have it working with current system.

And also as discussed in a few other threads. Headers only for auth is a no go as the only way as it's not compatible with casting in many cases like UPnP and a few other cases.

**Reply by x1arch** [reply]
Just think, if every a few month someone makes discussion about new API version, because old is outdated - maybe this is reason do not try reanimate the corpse and make a new standard?

**Reply by Tolriq** [reply]
Or maybe you can take 2 seconds to read the history and try to understand that you do not move dozen of servers and multiple dozens of client just like that ...
I've just checked your Github history about Subsonic and things like https://github.com/tamland/airsonic-refix/pull/108#issuecomment-1704433919 are enough to understand the situation.

**Comment by x1arch** [comment]
Ok, @Tolriq, @dweymouth, I understand your opinion.
Is anyone interested in the new API? Someone wants (ready) starts working on that?

**Reply by Tolriq** [reply]
Ha the innocence of youth :) You discovered Subsonic API 3 weeks ago and already think you know best and can move a dinosaur.
Already done a fork of @tamland project because he did not move fast enough and did not even bothered to do a proper fork to keep history :)
Good luck in your adventures, but this project is not the place to build your new soon dropped project, you can ping people individually on all the places as we did to build this project ...

**Reply by x1arch** [reply]
Is this is community or your private project? Why I can't discuss it with another peoples of community? In the head of repo writes Open, not ToIriq project. I hear you, thank you for your opinion, it's very important for me. :)

**Reply by Tolriq** [reply]
This is a project with rules and votes and it's already defined.

Coming here and try to gather people to start another competing project is as disrespectful as what you did with Tamland.

You are free to start your own project, but if I fork your front end, would you appreciate that I post there hey come to my fork it's better .....

So be respectful to what was build or go somewhere else and do better.

**Reply by x1arch** [reply]
Why you afraid? You have a history :), you have a many clients / servers which support this API. Is it will bad if some servers will support not one, two or more API?

PS. If you can to do this better, please do this, you can clone any open repository which you want, if license it's allowed. I have a plans for make this client is better and I do that. On the main page on the first line writes about project which forked. Whats wrong? Why you so angry?

**Reply by Tolriq** [reply]
Lol I guess you were never taught what respect is.

Unfortunately the code of conduct was not yet merged, not that you would have cared anyway.
